azure-proto-core/Resources/SystemAssignedIdentity.cs

```diff
@@ -0,0 +1,116 @@
+﻿namespace azure_proto_core.Resources
+{
+    using System;
+    using System.Text.Json;
+    using Azure.Core;
+
+    public class SystemAssignedIdentity
+    {
+        public Guid? TenantId { get; set; }
+
+        public Guid? PrincipalId { get; set; }
+
+        public SystemAssignedIdentity() { }
+
+        public SystemAssignedIdentity(Guid tenantId, Guid principalId)
+        {
+            TenantId = tenantId;
+            PrincipalId = principalId;
+        }
+
+        public int CompareTo(SystemAssignedIdentity other)
+        {
+            if (other == null)
+                return 1;
+
+            int compareResult = 0;
+            if ((compareResult = PrincipalId.GetValueOrDefault().CompareTo(other.PrincipalId.GetValueOrDefault())) == 0 &&
```

Comment:
if `this.PrincipalId != other.PrincipalId`, this will return the value of `this.PrincipalId.CompareTo(other.PrincipalId)`, due to short circuiting.  Is this the intention? (That is, the algorithm first compares principalId and only compares TenantId if PrincipalIds are equal.  If we think about using this for sorting, I think, sense Tenant is the container for ResourceId, this should be the other way around (that is Principals from the same Tenant should sort together).